### PR TITLE
Add tag functionality (Issue #508)

### PR DESCRIPTION
**Pull request by xvallspl**
Very early state, just testing the waters to gather feedback.

Add tag functionality (addresses issue #508)

Work on this PR:
- [x] Display tags in entry preview
- [x] Allow modification of tags on entry edit
- [x] Introduce search term "tag:"
- [x]  Implement new side panel, similar to the groups one, with the list of available tags. When clicking on one of those tags, the entry view widget should filter all entries that don't contain the selected tag.
- [x] Display tags as "pills"
- [x] Make tag editing an autocomplete dropdown
- [x] Adapt the styles, make everything beautiful (did the best I could)
- [x] Test 

Work for subsequent PRs:
- Update the menus with tag actions (still thinking which ones and what actions)
- Delete a tag from the database from the side panel
- When selecting a tag from the side panel filter groups and display only the ones containing entries with the tag
- Check feature/modifications suggestions in comments and Issue #508

## Screenshots

Screenshot 2021-11-24 at 19 57 50

Screenshot 2021-11-24 at 19 54 16

TagsBox2

## Testing strategy
Manually for the fast prototype, will come up with tests next.

## Type of change

- ✅ New feature (change that adds functionality)
- ✅ Documentation (non-code change)

### CODE REVIEW

**Comment by michaelk83** [comment]
> Figure out a way to improve tag editing and display

1. Tags edit box should be an auto-complete drop-down. IOW, you can:
    1. Click drop-down arrow to get a list of existing tags, select from the list to add a tag. Maybe with checkboxes to add/remove multiple?
    1. Type partial tag name to filter the drop-down list.
    1. If entered text does not exactly match an existing tag, list shows an additional entry: "New: [entered text]". Click that or press `Enter` to create a new tag with that name and add it to the entry.
    1. If entered text matches an existing tag, click that or press `Enter` to add the existing tag.
    1. Added tags become pills in the edit box. You can type the next tag after the pills.
    1. X on the right end of each tag pill allows to remove that tag.
1. In entry preview, tags show as pills near the bottom (no X's if the preview is not editable).
1. In tags side panel, can delete the selected tag with `Delete` key (with confirmation), or via menu. This only removes the tag from the entries and from the known tags list. The entries themselves remain.

You can have a look at GMail's labels UI for more ideas. There is probably a ready UI component for that autocomplete drop-down with pills. No need to reinvent the wheel here.

**Comment by xvallspl** [comment]
Added a side panel listing the tags in the active group. Double clicking on one of them filters by that tag in the entry view. 

Doubts/TODOs:

- Not sure double clicking is better than clicking
- No way to undo filtering yet
- Do we only want to show the tags from the current group or all of them?

**Comment by michaelk83** [comment]
My personal opinions:

> Not sure double clicking is better than clicking

It should behave the same as group selection.
 
> No way to undo filtering yet

Add an *(all)* meta tag at the top of the list in the side panel?
 
> Do we only want to show the tags from the current group or all of them?

I'd say, filter the groups to show only those that have matching entries, and in the preview panel list the entries in the selected group out the filtered ones.

**Comment by droidmonkey** [comment]
Concur with that, MVP for this PR

**Comment by xvallspl** [comment]
Added a couple more features and improvements: improved side panel, restore side panel selection after entry editing, tag pills view and autocomplete in Edit entry view.

The Tag Widget I used for the last two, https://github.com/nicktrandafil/tags, pulls code from QtGui/private, which either we add as a dependency or I can try to reimplement the functions we use. For the moment it makes the tests fail.

Looking right now at the formatting mistakes not picked up when running make format.

Updated the screenshots.

**Comment by michaelk83** [comment]
Looks like the auto-complete delay is pretty high. Is that a setting or a limitation of the code?  
I was thinking it should show and update the auto-complete list while you type.

**Comment by michaelk83** [comment]
There's a more official (?) fork at https://github.com/Qt-Widgets/tags . It says "3 commits ahead, 4 commits behind nicktrandafil:master". It might have fewer build issues?

**Comment by xvallspl** [comment]
> There's a more official (?) fork at https://github.com/Qt-Widgets/tags . It says "3 commits ahead, 4 commits behind nicktrandafil:master". It might have fewer build issues?

It has the same problem: #include <QtGui/private/qinputcontrol_p.h>

**Comment by xvallspl** [comment]
> Looks like the auto-complete delay is pretty high. Is that a setting or a limitation of the code?
> I was thinking it should show and update the auto-complete list while you type.

There's no delay, it just autocompletes (doesn't suggest by default) and my choice in the video was starting with a letter (case sensitive) that no other tag started with yet.

**Comment by droidmonkey** [comment]
Damn @xvallspl that is some great progress! The tag widget works well, but I agree it really shouldn't be using private functions.

**Comment by droidmonkey** [comment]
I'm pretty damn excited about this feature

**Comment by xvallspl** [comment]
@droidmonkey TeamCity is complaining about features I'm using from C++17 and C++14. Should we configure it for at least C++17? Should I stick to earlier standards?

**Comment by droidmonkey** [comment]
We are currently on C++ 14 standard. I think we can move to 17 though because xenial is out. @phoerious ?

**Comment by phoerious** [comment]
I guess.

**Comment by droidmonkey** [comment]
Although I noticed you are using std functions instead doing Qt functions to do some stuff with lists / sets. Favor Qt where possible.

**Comment by xvallspl** [comment]
Will do. I'm trying to understand how to implement multiline display for the tag widget when there are too many tags, but I am clueless so far.

**Comment by xvallspl** [comment]
I'm talking with the original author of the tags widget, @nicktrandafil, who is implementing another widget for multiline tags in https://github.com/nicktrandafil/tags/tree/tags_area. 

I'm waiting for a couple more features to be developed (scrolling, navigation) and I will adapt the widget to our needs. In the meanwhile, I will try to port the modifications I made (size of the rectangles, read only view) to his repository.

**Comment by michaelk83** [comment]
Do we really need the multi-line though? How many tags do we expect people to have? As a practical example, in my GMail, I never have more than 2-3 tags per email.

Assuming we're talking about the edit, I think as long as you can navigate left and right with the keyboard, a single line should be enough. When you select the edit, if there are too many tags, it can automatically scroll right to reveal the empty area for typing.

We can change it to a multi-line at a later point, if there's demand.

**Comment by michaelk83** [comment]
What I would consider changing, is in the preview pane, move the "Tags" section to near the bottom of the left column, and have it span both columns. The notes can be under that, and should probably span both columns too, since they can be long.

My original idea was to not have a "Tags" header at all in the preview, and just use pills, aligned to the bottom of the preview (similar to the edit, but not editable). But I'm not sure how that should work with long notes, and in any case, that can be left to a future PR.

**Comment by droidmonkey** [comment]
I tend to agree on the multi line edit, single line should be enough for now we can migrate in the future if folks are wanting it.

**Comment by xvallspl** [comment]
I do think multiline tags should be the default. In the end, if you don't have enough tags they will all be shown in a single line. and If you resize the window you probably want to be able to look at the tags at a single glance instead of having to navigate through them one by one seeing them scroll. If you don't want to wait, I could leave it for this PR though and implement the multicolumn tag span in the tag preview to have more space.

**Comment by michaelk83** [comment]
My impression from the screen shots is that the edit is wide enough to comfortably fit a decent number of tags on a single line. Most people probably won't have that many of them per entry, and won't be resizing the window very narrow. The few exceptions who might have too many tags can always resize the window wider - or post a feature request for a multi-line edit, if that really becomes an issue.

Let's not overthink things. Better to release early (as long as it's not broken) and gather actual user feedback from real-world use.

In the preview pane, I do think the tags list should span both columns. As it is, it looks rather narrow, and eats into the notes' space.

**Comment by xvallspl** [comment]
I'm still on this, just trying to find time and working the changes upstream on the plugin I'm using for the tags.

**Comment by droidmonkey** [comment]
@xvallspl this is awesome!

**Comment by droidmonkey** [comment]
One bit a feedback, it looks like the tag list filters based on the current group. That doesn't really make much sense since tags are supposed to be sort of meta groups that cross-cut your database. Prefer to have all known tags in the database be visible in the pick list at all times.

Another good thing to do would be to actually just populate the search box with the "tags:&lt;clicked tag&gt;" and let the search happen naturally from there. This has the benefit of showing the user that tags == search and it also lets the user narrow down the search directly while retaining the tag portion.

File: src/gui/tag/TagModel.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    tagList = s.toList();
+    tagList.sort();
+    tagList[0] = "All"; // Empty string always at the beginning
+    endResetModel();
+}
+
+int TagModel::rowCount(const QModelIndex& parent) const
+{
+    (void)parent; // avoid unused warnings
```
Comment (droidmonkey) [inline src/gui/tag/TagModel.cpp]: Use Q_UNUSED(x)

**Comment by xvallspl** [comment]
I was thinking of tags as confined to a group, just another way to group your entries within the group. I'll take a look at it as soon as I get out of the git applying hell I'm on.

**Comment by acolomb** [comment]
I strongly agree that tags should be global, not confined to a group. The purpose is to have a second grouping scheme based on tags, orthogonal to the groups scheme.

**Comment by droidmonkey** [comment]
I squashed all the intermediate commits and updated the translations and code format. One bug I found was that when adding a new tag to an entry, it didn't show up in the tag selection panel in the lower left until I locked/unlocked the database.

**Comment by xvallspl** [comment]
Thanks, Jonathan.  I'll take a look as soon as I can.

**Comment by eholzbach** [comment]
> I strongly agree that tags should be global, not confined to a group. The purpose is to have a second grouping scheme based on tags, orthogonal to the groups scheme.

This is what I've been hoping for. For example, I upgraded my primary Yubikey. I did not have a single place to track where it was being used. If I was able to tag entries in KeePassXC, it would be easy to list all the sites I need to update, without having to break my group schema.

**Comment by droidmonkey** [comment]
I picked this up and made substantial changes to implement the rest of the features. I even tossed in a couple default "tags" (more like saved searches) like expired entries and weak passwords. 

image

**Comment by xvallspl** [comment]
The sidebar looks so much better! Nice to see the changes in the searchbar!

**Comment by droidmonkey** [comment]
Finished up the code cleanup, this is ready for merge and 2.7.0!

**Review by droidmonkey — APPROVED** [review]

**Pull request merged**